src/ASM/GlbL2projector.C

```diff
@@ -33,6 +34,102 @@ LinAlg::MatrixType GlbL2::MatrixType   = LinAlg::SPARSE;
 LinSolParams*      GlbL2::SolverParams = nullptr;
 
 
+static void expandTensorGrid2 (const RealArray* in, RealArray* out)
+{
+  out[0].resize(in[0].size()*in[1].size());
+  out[1].resize(in[0].size()*in[1].size());
+
+  size_t i, j, ip = 0;
+  for (j = 0; j < in[1].size(); j++)
+    for (i = 0; i < in[0].size(); i++, ip++) {
+      out[0][ip] = in[0][i];
+      out[1][ip] = in[1][j];
+    }
+}
+
+
+static void expandTensorGrid3 (const RealArray* in, RealArray* out)
+{
+  out[0].resize(in[0].size()*in[1].size()*in[2].size());
+  out[1].resize(in[0].size()*in[1].size()*in[2].size());
+  out[2].resize(in[0].size()*in[1].size()*in[2].size());
+
+  size_t i, j, k, ip = 0;
+  for (k = 0; k < in[2].size(); k++)
+    for (j = 0; j < in[1].size(); j++)
+      for (i = 0; i < in[0].size(); i++, ip++) {
+        out[0][ip] = in[0][i];
+        out[1][ip] = in[1][j];
+        out[2][ip] = in[2][k];
+      }
+}
+
+
+L2ProbIntegrand::L2ProbIntegrand (const ASMbase& ASM, const IntegrandBase& itg) :
+  L2Integrand(ASM), m_itg(itg)
+{
+}
+
+
+bool L2ProbIntegrand::evaluate (Matrix& sField, const RealArray* gpar) const
+{
+  return m_asm.evalSolution(sField, m_itg, gpar);
+}
+
+
+int L2ProbIntegrand::dim () const
+{
+  return m_itg.getNoFields(2);
+}
+
+
+L2FuncIntegrand::L2FuncIntegrand (const ASMbase& ASM, const FunctionBase& func) :
+  L2Integrand(ASM), m_func(func)
+{
+}
+
+
+bool L2FuncIntegrand::evaluate (Matrix& sField, const RealArray* gpar) const
+{
+  std::array<RealArray,3> expanded;
+  if (dynamic_cast<const ASMunstruct*>(&m_asm) == nullptr) {
+    if (m_asm.getNoSpaceDim() == 2) {
+      expandTensorGrid2(gpar, expanded.data());
+      gpar = expanded.data();
+    } else if (m_asm.getNoSpaceDim() == 3) {
+      expandTensorGrid3(gpar, expanded.data());
+      gpar = expanded.data();
+    }
+  }
+
+  sField.resize(m_func.dim(), gpar[0].size());
+
+  Real2DMat u;
+  m_asm.getParameterDomain(u);
+  for (size_t i = 0; i < gpar[0].size(); ++i) {
+    Vec3 X;
+    std::vector<double> xi(3), param(3);
+    xi[0] = u[0][0] + gpar[0][i] / (u[0][1] - u[0][0]);
+    if (m_asm.getNoSpaceDim() > 1)
+      xi[1] = u[1][0] + gpar[1][i] / (u[1][1] - u[1][0]);
+    if (m_asm.getNoSpaceDim() > 2)
+      xi[2] = u[2][0] + gpar[2][i] / (u[2][1] - u[2][0]);
+    m_asm.evalPoint(xi.data(), param.data(), X);
+    std::vector<Real> vals = m_func.getValue(X);
+    for (size_t j = 1; j <= m_func.dim(); ++j)
+      sField(j,i+1) = vals[j-1];
+  }
+
+  return true;
+}
+
+
+int L2FuncIntegrand::dim () const
```

Comment:
`size_t`as return type? (since FunctionBase::dim() returns size_t)

Reply:
changed to size_t